### PR TITLE
Fixed == to require same-type arguments

### PR DESCRIPTION
**Pull request by tpoterba**
Assigning Dan because it involves a somewhat weird expr motif -- putting in a special comparison for typecheck, but using the function registry for the eval/compile

### CODE REVIEW

**Comment by tpoterba** [comment]
Fixes 1603

**Comment by tpoterba** [comment]
This actually caught a mistake in the docs:
```
[23:06:25][:makeHailDocs]     is.hail.utils.HailException: Cannot compare arguments of type Struct and String
[23:06:25][:makeHailDocs]     <input>:1:va.hweCase = gs.filter(g => sa.pheno == "Case").hardyWeinberg(),va.hweControl = gs.filter(g => sa.pheno == "Control").hardyWeinberg()
```

**Comment by danking** [comment]
@tpoterba I was expecting we'd change the function registry to require the same type variable for left and right, i.e.:

```scala
  register("==", (a: Any, b: Any) => a == b, null)(TTHr, TTHr, boolHr)
  register("!=", (a: Any, b: Any) => a != b, null)(TTHr, TTHr, boolHr)
```

Did that not work correctly?

**Comment by tpoterba** [comment]
No. You have to be able to compare ints and doubles. Also, we want attributes not to matter in struct comparison here.

**Comment by danking** [comment]
@tpoterba I thought the conversions would lift `Int`s to `Double`s before unifying the type variables. What are struct attributes? I am a unsure that our conversions would work for struct field types.

**Comment by tpoterba** [comment]
struct attrs meaning the info field metadata and such. These are included in type equality.

**Comment by danking** [comment]
Yeah I guess I'm OK with this, but it feels so wrong that `Field`s have metadata at the level of the type that is relevant to unification and equality. I guess even if that was fixed you wouldn't be able to compare `Struct`s whose fields had comparable but not equal types.

Yeah, this is a necessary evil.

File: src/test/scala/is/hail/methods/ExprSuite.scala [hunk 1]
```diff
@@ -833,6 +833,19 @@ class ExprSuite extends SparkSuite {
       assert(eval("-2 ** -2").contains(-0.25))
     }
 
+    assert(eval("1 == 1.0").contains(true))
+    assert(eval("[1,2] == [1.0, 2.0]").contains(true))
+    assert(eval("[1,2] != [1.1, 2.0]").contains(true))
+    assert(eval("{a: 1, b: NA: Genotype} != {a: 2, b: NA: Genotype}").contains(true))
+    assert(eval("{a: 1, b: NA: Genotype} == {a: 1, b: NA: Genotype}").contains(true))
+
+    //    TestUtils.interceptFatal("cannot compare")
```
Comment (danking) [inline src/test/scala/is/hail/methods/ExprSuite.scala]: Remove comment

File: src/main/scala/is/hail/expr/Type.scala [hunk 2]
```diff
@@ -523,6 +534,12 @@ case class TSet(elementType: Type) extends TIterable {
 
 case class TDict(keyType: Type, valueType: Type) extends TContainer {
 
+  override def canCompare(other: Type): Boolean = other match {
+    case TDict(okt, ovt) => keyType.canCompare(okt) && valueType.canCompare(ovt)
+    case _ => false
+  }
+
+
```
Comment (danking) [inline src/main/scala/is/hail/expr/Type.scala]: extra newline

**Comment by tpoterba** [comment]
addressed comments, back to you

**Pull request merged**